File: backend/Models/Quiz.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace prid.Models;

public enum Status
{
    EN_COURS = 0,
    FINI = 1,
    PAS_COMMENCE = 2,
    CLOTURE = 3
}

public class Quiz
{
    [Key]
    public int Id { get; set; }
    public string Name { get; set; } = null!;
    public string? Description { get; set; }
    public int DatabaseId { get; set; }
    public bool IsPublished { get; set; }
    public bool IsTest { get; set; }
    public DateTimeOffset? StartDate { get; set; }
    public DateTimeOffset? EndDate { get; set; }

    public ICollection<Question> Questions { get; set; } = new HashSet<Question>();
    public ICollection<Attempt> Attempts { get; set; } = new HashSet<Attempt>();
    public Database Database { get; set; } = null!;

    [NotMapped]
    public string? Status { get; set; }

    [NotMapped]
    public string? FirstQuestionId { get; set; }

    [NotMapped]
    public Attempt? LastAttempt { get; set; }

    [NotMapped]
    public bool? HasAttempt { get; set; }

    public Quiz() { }
}

File: frontend/src/app/models/quiz.ts
import { Type } from "class-transformer";
import { format } from "date-fns";
import { Database } from "./database";
import { Attempt } from "./attempts";
import { Question } from "./question";

export class Quiz {
    id?: number
    name?: string
    description?: string | null
    databaseId?: number
    isPublished?: boolean
    isTest?: boolean
    @Type(() => Date)
    startDate?: Date
    @Type(() => Date)
    endDate?: Date

    status?: string
    firstQuestionId?: string
    lastAttempt?: Attempt
    hasAttempt?: boolean

    database?: Database
    attempts: Attempt[] = []
    questions: Question[] = []

    get start(): string {
        return `${this.startDate ? format(this.startDate!, 'dd/MM/yyyy') : 'N/A'}`;
    }

    get end(): string {
        return `${this.endDate ? format(this.endDate!, 'dd/MM/yyyy') : 'N/A'}`;
    }

    get type(): string {
        return `${this.isTest ? 'Test' : 'Training'}`;
    }

    get databaseName(): string {
        return this.database!.name!
    }

    get evaluation(): string {
        if (this.attempts.length > 0) {
            const lastAttempt = this.attempts[this.attempts.length - 1];
            return `${lastAttempt.evaluation?.toString() + '/10'}`;
        } else {
            return 'N/A';
        }
    }
}